core/include/io.h

```diff
@@ -273,4 +273,37 @@ static inline void io_clrsetbits8(vaddr_t addr, uint8_t clear_mask,
 	io_write8(addr, (io_read8(addr) & ~clear_mask) | set_mask);
 }
 
+/*
+ * Poll on a IO memory content or timeout
+ *
+ * @addr is the address of the memory cell accessed
+ * @val represents the val  of the memory cell accessed
+ * @cond represents the condition to get the correct value
+ * @delay_us represents the read interval in mircorseconds
+ * @_timeout_us represents the time-out period in microseconds
+ *
+ * @return true means time-out, false means got right value
+ */
+#define IO_READ32_POLL_TIMEOUT(addr, val, cond, delay_us, _timeout_us) \
+	({ \
+		uint64_t _timeout = 0; \
+		uint64_t _delay_us = delay_us; \
+		bool _flag = true; \
+		\
+		while (_timeout < (_timeout_us)) { \
+			(val) = io_read32(addr); \
+			if (cond) { \
+				_flag = false; \
+				break; \
+			} \
+			_timeout += (_delay_us); \
+			udelay(_delay_us); \
+		} \
+		(val) = io_read32(addr); \
+		if (cond) { \
+			_flag = false; \
+		} \
```

Comment:
Could simply:
```c
#define IO_READ32_POLL_TIMEOUT(_addr, _val, _cond, _delay_us, _timeout_us) \
	({ \
		uint64_t __timeout = 0; \
		uint64_t __delay = (_delay_us); \
		\
		while (__timeout < (_timeout_us)) { \
			(_val) = io_read32(_addr); \
			if (_cond) \
				break; \
			__timeout += (__delay); \
			udelay(__delay); \
		} \
		(_val) = io_read32(_addr); \
		(_cond); \
	})

Reply:
I'm not sure we want to evaluate `_cond` more times than necessary.

Reply:
We need to evaluate `cond` even once timeout has elapsed: https://github.com/OP-TEE/optee_os/pull/6297#discussion_r1330511820.
Otherwise, it will be the caller responsibility to check again its condition in case the macro returns `true` and the caller executes in a theraded context with foreign interrupts unmasked.

Reply:
Got it, good point.

Reply:
> Could **~simply~ simplify**:
> 
> ```c
> #define IO_READ32_POLL_TIMEOUT(_addr, _val, _cond, _delay_us, _timeout_us) \
> 	({ \
> 		uint64_t __timeout = 0; \
> 		uint64_t __delay = (_delay_us); \
> 		\
> 		while (__timeout < (_timeout_us)) { \
> 			(_val) = io_read32(_addr); \
> 			if (_cond) \
> 				break; \
> 			__timeout += (__delay); \
> 			udelay(__delay); \
> 		} \
> 		(_val) = io_read32(_addr); \
> 		(_cond); \
> 	})
> ```

It's awesome, far better than the my version!!!